proposals/0016-simd-application-fees.md

```diff
@@ -0,0 +1,149 @@
+---
+simd: '0016'
+title: Application Fees (Write-lock fees)
+authors:
+  - Godmode Galactus (Mango Markets)
+  - Maximilian Schneider (Mango Markets)
+category: Fees
+type: Fees
+status: Draft
+created: 2022-12-23
+---
+
+## Problem
+
+According to the discussion on the following issue:
+https://github.com/solana-labs/solana/issues/21883
+
+During network congestion, many transactions in the network cannot be processed 
+because they all want to write lock similar accounts. 
+When a write lock on an account is taken by a transaction batch no other batch 
+can use this account in parallel, so only transactions belonging to a single batch 
+are processed correctly and all others are forwarded to the leader of the next slot. 
+There are multiple accounts of OpenBook (formerly serum), mango markets, etc which 
+are used in high-frequency trading. During the event of extreme congestion we observe 
+that specialized trading programs write lock these accounts but never CPI into the respecitve 
+programs unless they can extract profit, effectively starving the actual users for access. 
+Penalizing this behaviour with runtime fees collected by validators can create a perverse 
+incentive to artificially delay HFT transactions, cause them to fail and charge penalty fees.
+
+## Solution
+
+As a high-performance cluster, we want to incentivize players which feed the network 
+with correctly formed transactions instead of spamming the network. Introduction 
+of application fees would be interesting way to penalize the bad actors and applications 
+can rebate these fees to the good actors. This means the application has to decide who to 
+rebate and who to penalize through lamport transfers. In particular it needs to be able to 
```

Comment:
> `This means the application has to decide who to rebate and who to penalize through lamport transfers.`

How the application can reliably decide who to penalize _when bad actors' transactions does cpi good-actor looking ixes if they cannot extract profit after directly peeking the market state by its own?_

Reply:
It is more like application developers will have power to decide who are good and bad actors. A good actor will create a valid transaction, a bad actor will usually do transactions which will check market conditions and cancel the transaction if market conditions are not met.
So application developers can apply application fees for unintended use of the application.

Reply:
yeah, i understand the motivation and just impressed with your actual work including code.

however, can this be fraud proof without compromise of dx/comparability?

Specifically, I'm commenting here after skimming through the code, specifically: https://github.com/godmodegalactus/paper-clip-maximizer/blob/main/programs/paper-clip-maximizer/src/lib.rs#L113 and https://github.com/solana-labs/solana/pull/30137/files#r1097407270


consider this bot tx with 2 ixes: 

```
ApplicationFeeAvadingTx:
  IX1: directly looks market condition without cpi-ing target dapp at all and if this makes profit after tx fee + application fee:
       true => do the arb.
       false => no-op; fall through to the IX2

  IX2: if arb-ed at IX1: 
       true => no-op; (assuming the daap won't allow rebate because of executed arb)
       false => CPI so that this tx is indistinguishable from good actors's well-intended behavior:
                        *the daap will rebate.*
```

so, i think bot can bypass the application fee? how can the dapp detect _not to rebate the tx's application fee_? Maybe, we can resort to instructionssysvar or flashloan-like construct or get_sibling_ix syscall. However, there's some composability issue and i'm a bit worried that some loophole will be spotted eventually.

That's more detailed explanation of _reliably_ in my first comment: https://github.com/solana-foundation/solana-improvement-documents/pull/16#discussion_r1089665912

Reply:
instruction sysvar introspection is becoming the standard, turns out that limiting 3rd party programs is the only way how a dapp can effectively prevent aggressive use, a common pvp strategy in HFT is to cause exchange delay / overload.

Reply:
Mango v4 would be able to charge specifically users who spam risk-free arbitrage or spam liquidations by increasing the write-lock fee on perp markets, token banks and user mango accounts.

1. perp market:
    
    write-locks: perp_liq, perp_place, perp_cancel, perp_consume, perp_settle_fees, perp_update_funding
    
    rebate on:
    
    1. successful liquidation / consume_events
    2. HFT mm refresh (cancel all, n x place POST, no IOC)
2. token vaults: (could also be implemented on openbook market)
    
    write-locks: serum3_liq, deposit, withdraw
    
    rebate on:
    
    1. successful liquidation
    3. place IOC & fill in isolation
    4. HFT mm refresh (cancel all, n x place POST, no IOC)
3. mango account:
    
    write-locks: all liq tx, consume_events, settle_pnl, all user signed tx
    
    rebate on:
    
    1. tx signed by owner / delegate
    2. successful liquidation / settlement / consume_events
    

### Implementation:

1a/1b/2a/2b/3b — all these instructions are difficult to spam successfully. naively spamming bots have an error / success ratio of around 50/1. basically these instructions just need to call the rebate syscall at the end of the instruction passing the perp_market / token_vault / mango_accounts concerned.

1c/2b/2c/2d —  require explicit whitelisting of recommended behavior, they restrict composability for the sake of having a pleasant experience for good actors. a (simplified) record of all instructions executed inside the health region could be verified by the health_region_end instruction. this should be limited to <128 instructions to reduce the amount of account space consumed.

3a — very straightforward

Reply:
Thanks a lot,

You are right, it is possible to do something similar to avoid paying application fees. And it will be hard job for the DApp developers to effectively implement strategies to avoid similar kinds of attacks. I think this can be avoided if the DApp can either whitelist or blacklist instruction order in the transaction using the instruction sysvar. Or force the format of the transaction should respect certain norms to get the rebates